Automattic/woocommerce-payments
# Use the customer ID saved in the subscription to process renewal payments

Fixes #8892

#### Changes proposed in this Pull Request

When processing renewals using WooCommerce subscriptions, WooPayments will try to retrieve the Customer ID from the user meta. In most scenarios this is acceptable; however, we have received reports that after performing a migration, the Customer ID in the User meta may get lost, when this happens, a new Customer ID will be created, but this will cause an error since the Payment details stored in the subscription won't match the new Customer ID.

This PR will use the Customer ID saved in the subscription to make sure there are no mismatches between it and the payment details during renewals.

#### Testing instructions

* Install WooPayments and WooCommerce Subscriptions
* Set up a Subscription Product and purchase it. 
* In WP Admin, find the generated subscription. Note the User ID of the customer and the Subscription ID.
* In a mysql console (or using phpMyAdmin) run the following commands, replace the USER_ID and SUBSCRIPTION_ID:
```sql
SELECT * FROM `wp_usermeta` WHERE user_id = <USER_ID> AND meta_key LIKE 'wp__wcpay%';

SELECT * FROM `wp_wc_orders_meta` WHERE meta_key LIKE '%customer_id' AND order_id = <SUBSCRIPTION_ID>;
```
* Both values should match.
* Run the following command to update the WooPayments Customer ID in the user meta:
```sql
UPDATE `wp_usermeta` SET `meta_value` = 'cus_INVALID' WHERE user_id = <USER_ID> AND meta_key LIKE 'wp__wcpay%' LIMIT 1;
``` 
* In WP Admin, open the Subscription details, on the right side find the Order Actions, and then select "Process Renewal'. Click "Update".

##### Before
* The payment of the renewal order fails. Note the renewal order ID, and run this SQL:
```sql
SELECT * FROM `wp_wc_orders_meta` WHERE meta_key LIKE '%customer_id' AND order_id = <RENEWAL_ORDER_ID>;
```
* The customer ID value should be: `cus_INVALID`

##### After
* The payment of the renewal order succeeds. Note the renewal order ID, and run this SQL:
```sql
SELECT * FROM `wp_wc_orders_meta` WHERE meta_key LIKE '%customer_id' AND order_id = <RENEWAL_ORDER_ID>;
```
* The customer ID value should be the same as the one in the Subscription.

-------------------


**Post merge**
